### PR TITLE
MAINT: refactor "for ... in range(len(" statements

### PR DESCRIPTION
**Pull request by mwtoews**
Refactor `for i in range(len(items))` to `for item in items`, which is generally regarded to be more Pythonic.

Other instances refactored to use either `zip` or `enumerate`.

Cases were identified using `git grep " in range(len("`, although many other cases not suitable for refactoring.

### CODE REVIEW

**Comment by mattip** [comment]
There are three patterns of replacement:
- replace `for i in range(len(x)):` with `for item in x:`
- replace `for i in range(len(x)):` with `for i, item in enumerate(x):`
- replace `for i in range(len(x)):` with `for itemx, itemy in zip(x, y):`

I think the first one is a clear win. I am not so excited about the other two. Is there a clear preference in our style guide? I think I would prefer to treat them as we do other code cleanups: only touch them if we need to do so for other reasons.

**Comment by eric-wieser** [comment]
IMO all three are overall wins, but won't comment on whether the wins are strong enough to justify the review time. From a quick skim I think this is some nice cleanup.

**Comment by rkern** [comment]
When replacing with `zip()`, especially in the test utilities, be sure to handle differing lengths correctly.

**Comment by mwtoews** [comment]
The motivation behind this PR is part style and performance. The `for` statement in Python is really a [foreach](https://en.wikipedia.org/wiki/Foreach_loop) statement, so it's not the same as C. Iterating over the simplest statements yields the best performance when scaled up (using Python 3.9.6):
```python
ar = np.arange(10000, dtype=float)

%timeit for x in ar: _ = x
# 334 µs ± 752 ns per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit for i, x in enumerate(ar): _ = x
# 552 µs ± 10.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit for a, b in zip(ar, ar): _, _ = a, b
# 730 µs ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
%timeit for i in range(len(ar)): _ = ar[i]
# 746 µs ± 1.93 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Where there are shorter iterations, the performance is about the same for all methods.

Mark any edits that are _less_ readable, and I'll cull them.

**Comment by mwtoews** [comment]
> When replacing with `zip()`, especially in the test utilities, be sure to handle differing lengths correctly.

Thanks, I was paying attention to this aspect, as zip will stop on the shortest iteration. But also `x[i], y[i]` would have raised `IndexError` before if they were different lengths.

File: numpy/core/records.py [hunk 1]
```diff
@@ -673,8 +673,8 @@ def fromarrays(arrayList, dtype=None, shape=None, formats=None,
     _array = recarray(shape, descr)
 
     # populate the record array (makes a copy)
-    for i in range(len(arrayList)):
-        _array[_names[i]] = arrayList[i]
+    for name, item in zip(_names, arrayList):
+        _array[name] = item
```
Comment (mattip) [inline numpy/core/records.py:677]: Could you merge the check in lines 667-671 into this loop so we iterate once over `arrayList`

File: numpy/compat/_inspect.py [hunk 2]
```diff
@@ -154,10 +154,10 @@ def formatargspec(args, varargs=None, varkw=None, defaults=None,
...
     if defaults:
         firstdefault = len(args) - len(defaults)
-    for i in range(len(args)):
-        spec = strseq(args[i], formatarg, join)
-        if defaults and i >= firstdefault:
-            spec = spec + formatvalue(defaults[i - firstdefault])
+    for idx, arg in enumerate(args):
+        spec = strseq(arg, formatarg, join)
+        if defaults and idx >= firstdefault:
+            spec += formatvalue(defaults[idx - firstdefault])
```
Comment (mattip) [inline numpy/compat/_inspect.py:160]: I prefer the original code here

File: numpy/core/tests/test_umath_complex.py [hunk 3]
```diff
@@ -134,8 +134,8 @@ def test_simple(self):
         x = np.array([1+0j, 1+2j])
         y_r = np.log(np.abs(x)) + 1j * np.angle(x)
         y = np.log(x)
-        for i in range(len(x)):
-            assert_almost_equal(y[i], y_r[i])
+        for yi, y_ri in zip(y, y_r):
+            assert_almost_equal(yi, y_ri)
```
Comment (mattip) [inline numpy/core/tests/test_umath_complex.py:138]: Couldn't this be written more compactly without the loop as `assert_almost_equal(y, y_r)`? What am I missing?

File: numpy/distutils/misc_util.py [hunk 4]
```diff
@@ -113,10 +113,9 @@ def quote_args(args):
     # don't used _nt_quote_args as it does not check if
     # args items already have quotes or not.
     args = list(args)
-    for i in range(len(args)):
-        a = args[i]
-        if ' ' in a and a[0] not in '"\'':
-            args[i] = '"%s"' % (a)
+    for idx, arg in enumerate(args):
+        if ' ' in arg and arg[0] not in '"\'':
+            args[idx] = f'"{arg}"'
```
Comment (mattip) [inline numpy/distutils/misc_util.py:118]: I don't see anywhere in numpy or scipy that uses this function. Can we deprecate it?

File: numpy/f2py/f2py2e.py [hunk 5]
```diff
@@ -359,33 +359,34 @@ def buildmodules(lst):
     cfuncs.buildcfuncs()
     outmess('Building modules...\n')
     modules, mnames, isusedby = [], [], {}
-    for i in range(len(lst)):
-        if '__user__' in lst[i]['name']:
-            cb_rules.buildcallbacks(lst[i])
+    for item in lst:
+        if '__user__' in item['name']:
+            cb_rules.buildcallbacks(item)
```
Comment (mattip) [inline numpy/f2py/f2py2e.py:364]: The changes here look like a nice cleanup

File: numpy/lib/polynomial.py [hunk 6]
```diff
@@ -1273,14 +1272,14 @@ def fmt_float(q):
...
+        for k, coeff in enumerate(coeffs):
+            if not iscomplex(coeff):
+                coefstr = fmt_float(real(coeff))
+            elif real(coeff) == 0:
+                coefstr = '%sj' % fmt_float(imag(coeff))
             else:
-                coefstr = '(%s + %sj)' % (fmt_float(real(coeffs[k])),
-                                          fmt_float(imag(coeffs[k])))
+                coefstr = '(%s + %sj)' % (fmt_float(real(coeff)),
+                                          fmt_float(imag(coeff)))
```
Comment (mattip) [inline numpy/lib/polynomial.py:1282]: This looks better now

**Comment by mattip** [comment]
OK, I had a deeper look. Only one of these looks out of place, the rest are either improvements or "doesn't matter". Some of the code can be cleaned up further.

**Comment by rkern** [comment]
> Thanks, I was paying attention to this aspect, as zip will stop on the shortest iteration. But also `x[i], y[i]` would have raised `IndexError` before if they were different lengths.

Yes, and that needs to continue to be the case, especially for the functions in the tests. That `IndexError` is part of the test.

File: numpy/compat/_inspect.py (hunk 2 above)
Comment (mwtoews) [reply numpy/compat/_inspect.py]: Change backed out.

File: numpy/core/records.py (hunk 1 above)
Comment (mwtoews) [reply numpy/core/records.py]: Makes sense, done.

File: numpy/core/tests/test_umath_complex.py (hunk 3 above)
Comment (mwtoews) [reply numpy/core/tests/test_umath_complex.py]: Must have been an oversight, as `assert_almost_equal` takes array_like args, and seems to work with complex types. Done 3x.

File: numpy/distutils/misc_util.py (hunk 4 above)
Comment (mwtoews) [reply numpy/distutils/misc_util.py]: I've backed out this change, and will follow-up in a separate issue/PR later.

**Review by mwtoews — COMMENTED** [review]
@mattip thanks, comments addressed in updated commit (via `push -f`)

File: numpy/lib/tests/test_function_base.py [hunk 7]
```diff
@@ -2773,8 +2773,8 @@ def test_period(self):
 
 
 def compare_results(res, desired):
-    for i in range(len(desired)):
-        assert_array_equal(res[i], desired[i])
+    for result_item, desired_item in zip(res, desired):
+        assert_array_equal(result_item, desired_item)
```
Comment (mattip) [inline numpy/lib/tests/test_function_base.py:2777]: Does this need a `if len(res) != len(desired): raise ...` or is it guarenteed that they will be equal length in all the tests?

File: numpy/lib/tests/test_shape_base.py [hunk 8]
```diff
@@ -713,5 +713,5 @@ def test_basic(self):
 
 # Utility
 def compare_results(res, desired):
-    for i in range(len(desired)):
-        assert_array_equal(res[i], desired[i])
+    for result_item, desired_item in zip(res, desired):
+        assert_array_equal(result_item, desired_item)
```
Comment (mattip) [inline numpy/lib/tests/test_shape_base.py:717]: Does this need a if len(res) != len(desired): raise ... or is it guarenteed that they will be equal length in all the tests?

**Comment by mattip** [comment]
This looks good now, just two small nits that if broken now they were broken before too.

File: numpy/lib/tests/test_function_base.py (hunk 7 above)
Comment (mwtoews) [reply numpy/lib/tests/test_function_base.py]: It seems this function is not used, so it is now removed.

File: numpy/lib/tests/test_shape_base.py (hunk 8 above)
Comment (mwtoews) [reply numpy/lib/tests/test_shape_base.py]: I've added a length check, with a comment for [PEP 618](https://www.python.org/dev/peps/pep-0618/) for Python 3.10, when these functions should be refactored to `zip(..., strict=True)`.

It also turns out there was one failure (`test_integer_split_2D_rows`), where the lengths were not equal, which is simple to fix in this PR.

**Comment by mwtoews** [comment]
@mattip the main changes to review are in `numpy/lib/tests/test_shape_base.py`

File: numpy/lib/tests/test_shape_base.py [hunk 9]
```diff
@@ -392,7 +392,7 @@ def test_integer_split_2D_rows(self):
         assert_(a.dtype.type is res[-1].dtype.type)
 
         # Same thing for manual splits:
-        res = array_split(a, [0, 1, 2], axis=0)
+        res = array_split(a, [0, 1], axis=0)
         tgt = [np.zeros((0, 10)), np.array([np.arange(10)]),
                np.array([np.arange(10)])]
         compare_results(res, tgt)
```
Comment (mwtoews) [inline numpy/lib/tests/test_shape_base.py:398]: the alternative change is to keep this `res`, and add a fourth array to `tgt`.

**Comment by mattip** [comment]
Thanks @mwtoews for working through all this

File: numpy/distutils/misc_util.py (hunk 4 above)
Comment (mwtoews) [reply numpy/distutils/misc_util.py]: xref #19811

**Pull request merged**